rspace/src/main/scala/coop/rchain/rspace/history/package.scala

```diff
@@ -201,27 +256,115 @@ package object history {
             // If the "tip" is an existing node, then we can add the new leaf's hash to the node's
             // pointer block and rehash.
             case Node(pb) =>
-              val pathLength    = parents.length
-              val newLeafIndex  = JByte.toUnsignedInt(encodedKeyNew(pathLength))
-              val hd            = Node(pb.updated(List((newLeafIndex, LeafPointer(newLeafHash)))))
+              val pathLength      = parents.countPathLength
+              val wholePathLength = encodedKeyNew.size
+              val pathLeft        = wholePathLength - pathLength
+              val newLeafIndex    = JByte.toUnsignedInt(encodedKeyNew(pathLength))
+              logger.debug(
+                s"Node state, current:$pathLength, left: $pathLeft, whole: $wholePathLength")
+              val ptr = if (pathLeft > 1) {
+                val skip = Skip(ByteVector(encodedKeyNew.splitAt(pathLength + 1)._2),
+                                LeafPointer(newLeafHash))
+                val skipHash = Trie.hash(skip)(codecK, codecV)
+                store.put(txn, skipHash, skip)
+                NodePointer(skipHash)
+              } else {
+                LeafPointer(newLeafHash)
+              }
+              val hd            = Node(pb.updated(List((newLeafIndex, ptr))))
               val rehashedNodes = rehash[K, V](hd, parents)
               val newRootHash   = insertTries(store, txn, rehashedNodes).get
               store.putRoot(txn, branch, newRootHash)
               logger.debug(s"workingRootHash: $newRootHash")
+            // If the tip is a skip node -> there is no Node for this Leaf.
+            // need to shorten this skip and rectify the structure
```

Comment:
When you say "shorten" does this also include removing it in the case where we the shorten it to 0?

Reply:
huh... good point. lemme explore that.

Reply:
yes, it means exactly that. I'll change the comment to reflect that better (it's the first case).